### PR TITLE
NT-Xent Contrastive Loss Implementation

### PR DESCRIPTION
**Pull request by JohnKala**
# NT-Xent Contrastive Loss Implementation

## Overview
This PR implements the NT-Xent (Normalized Temperature-scaled Cross Entropy) contrastive loss function in the BEAD framework. NT-Xent enables self-supervised contrastive learning capabilities, allowing models to learn better representations by contrasting positive pairs against negative pairs in the latent space.

## Key Features
- **New Loss Functions**: Added [NTXentLoss](cci:2://file:///Users/jokala/BEAD_upstream/bead/src/utils/loss.py:832:0-907:22) and [NTXentCombinedLoss](cci:2://file:///Users/jokala/BEAD_upstream/bead/src/utils/loss.py:910:0-996:59) classes
- **Latent Space Augmentation**: Created `ntxent_utils.py` with augmentation functions
- **Smart Unpacking**: Refactored model output handling to support contrastive learning
- **Configuration-Driven**: All NT-Xent capabilities are controlled via configuration parameters
- **Backward Compatible**: Existing models work without changes when NT-Xent is disabled

## Implementation Details
1. The implementation uses a combined loss pattern that wraps any existing loss function
2. Augmentations are applied to latent vectors using configurable noise transformations
3. Training pipeline modifications are minimal and only active when NT-Xent is enabled

## Testing
- Added unit tests for NT-Xent loss and helper functions
- Successfully tested end-to-end with `Planar_ConvVAE` model
- Additional model architectures have been tested offline

## Parameters Added
- `c.use_ntxent`: Boolean to enable/disable NT-Xent functionality
- `c.loss_function`: Set to "NTXentCombinedLoss" when using NT-Xent
- `c.base_loss_function`: Base loss to combine with NT-Xent (e.g., "VAEFlowLoss")
- `c.ntxent_temperature`: Temperature parameter for NT-Xent (default: 0.07)
- `c.ntxent_noise_sigma`: Noise sigma for augmentations (default: 0.1)
- `c.ntxent_weight`: Weight for NT-Xent loss component (default: 0.5)

## Future Work
- Add additional augmentation strategies beyond Gaussian noise
- Performance benchmarking against non-contrastive baselines
- Integration with additional model architectures

### CODE REVIEW

File: bead/src/models/models.py [hunk 1]
```diff
@@ -377,7 +377,7 @@ def forward(self, x):
         # Apply softmax to map z to simplex (Dirichlet approximation)
         D_z = torch.nn.functional.softmax(G_z,dim=-1)
         out = self.decode(D_z)
-        return out, mean, logvar, D_z,D_z,D_z
+        return out, mean, logvar, D_z
```
Comment (PRAkTIKal24) [inline bead/src/models/models.py:380]: pls check the current main after DVAE merge- The return here should have both `G_z` and `D_z`. Pls check for consistency with the DVAE losses (VAE and VAEFlow) too from current `main`.

File: bead/src/utils/helper.py [hunk 2]
```diff
@@ -924,6 +924,63 @@ def call_forward(model, inputs):
...
+    For models that don't naturally produce all these values:
+    - AE models (len(outputs) == 2): Returns (recon, zeros, zeros, zeros, z, z)
+    - VAE models (len(outputs) == 4): Returns (recon, mu, logvar, zeros, z, z)
+    - Flow models (len(outputs) == 6): Returns outputs as-is
+    
+    Args:
+        outputs (tuple): The raw outputs tuple from a model's forward method.
+            Could be one of:
+            - (recon, z) for basic autoencoders (AE, AE_Dropout_BN, ConvAE)
+            - (recon, mu, logvar, z) for VAEs (ConvVAE, Dirichlet_ConvVAE)
```
Comment (PRAkTIKal24) [inline bead/src/utils/helper.py:943]: DVAE should have `G_z` and `D_z` as part of the return statement - you could call these `z0`, `zk` for consistency or not, you choose :)

File: bead/src/utils/loss.py [hunk 3]
```diff
@@ -126,7 +126,7 @@ def calculate(self, recon, target, mu, logvar, parameters=None, log_det_jacobian
 
         else:
             kl_loss = -0.5 * torch.sum(1 + logvar - mu.pow(2) - logvar.exp(), dim=1)
-            return (kl_loss/batch_size,)
+            return (kl_loss.mean(),)
```
Comment (PRAkTIKal24) [inline bead/src/utils/loss.py:129]: we don't want `mean()` here because the `kl_loss` above is already summed across the batch

File: bead/src/utils/loss.py [hunk 4]
```diff
@@ -62,7 +62,7 @@ def __init__(self, config):
         self.reg_param = config.reg_param
         self.component_names = ["reco"]
 
-    def calculate(self, recon, target, mu, logvar, parameters, log_det_jacobian=0):
+    def calculate(self, recon, target, mu, logvar, parameters, log_det_jacobian=0, zk=None, generator_labels=None, **kwargs):
```
Comment (PRAkTIKal24) [inline bead/src/utils/loss.py:65]: not sure what these extra arguments are doing here when they're not used in the call? `**kwargs` is fine to make the `calculate()`  consistent across functions that require more arguments but you wanna have a single call in say `training.py`. 

Even in that case the individual loss classes are not intended to be used in training- we will always use some combined loss class (eg. VAELoss, VAEFlowLoss, VAESupConLoss etc.) - so as long as the `calculate()` definitions in these are consistent in terms of the call arguments, we're fine. 

You could consider replacing things like `generator_labels` in losses like `VAELoss` which don't use them with `**kwargs` so that the call in `training.py` is invariant to the loss function chosen

File: bead/src/trainers/training.py [hunk 5]
```diff
@@ -97,21 +97,34 @@ def fit(
...
+                zk_j = None  # No second view for standard training
+
+            # Prepare common arguments for loss calculation
+            loss_args = {
+                "recon": recon,
+                "target": inputs,
+                "mu": mu,
+                "logvar": logvar,
+                "zk": zk,
+                "z0": z0,  # Include z0 for flow-based loss functions
```
Comment (PRAkTIKal24) [inline bead/src/trainers/training.py:116]: we don't really use `z0, zk` in the flow loss calculations - SupCon currently is the only one that uses `zk`. Does nt-xent use `z0`? if not, could consider simplifying this argument

File: bead/src/trainers/training.py [hunk 6]
```diff
@@ -205,20 +218,34 @@ def validate(
...
-                    logvar=logvar,
-                    zk=zk,
-                    parameters=model_for_loss_params.parameters(),
-                    log_det_jacobian=ldj
-                    if hasattr(ldj, "item")
-                    else torch.tensor(0.0, device=device),
-                    generator_labels=gen_labels,
-                )
+                # If NT-Xent is enabled, use dual forward pass with augmented views
+                if hasattr(config, "use_ntxent") and config.use_ntxent:
```
Comment (PRAkTIKal24) [inline bead/src/trainers/training.py:222]: the flag should just use the name of the loss and check if nxent is present in the string to decide whether we need the second ntxent pass as discussed in the meeting :)

File: bead/src/utils/ggl.py [hunk 7]
```diff
@@ -191,6 +191,13 @@ class Config:
     overlay_roc_projects: list
     overlay_roc_save_location: str
     overlay_roc_filename: str
+    
+    # NT-Xent loss parameters
+    use_ntxent: bool = False  # Whether to use NT-Xent contrastive loss
+    base_loss_function: str = "VAELoss"  # Base loss function to wrap with NT-Xent
```
Comment (PRAkTIKal24) [inline bead/src/utils/ggl.py:197]: above to config options are not required as discussed in the meeting

File: bead/src/utils/ntxent_utils.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+This module provides utility functions for implementing NT-Xent loss in BEAD models.
+Functions include data augmentation methods for creating contrastive views of inputs.
+"""
+
+import torch
+
+
+def generate_augmented_views(x, sigma=0.1):
+    """
+    Generate two augmented views of the input by adding Gaussian noise.
```
Comment (PRAkTIKal24) [inline bead/src/utils/ntxent_utils.py:13]: for consistency with particle physics lingo let's call this " applying a naiive gaussian smearing strategy" as opposed  to "adding gaussian noise", so that the docs website will have a searchable smearing tag

File: tests/unit/test_ntxent_utils.py
Comment (PRAkTIKal24) [inline tests/unit/test_ntxent_utils.py:1]: great job on adding unit tests on the first pass! - however, can we convert the test names to `test_helper`, and `test_loss`. also move the ntxent_utils test into `test_helper` - just so the github CI is cleaner. in the future we coud add other helper/loss unit tests into the same respective files

**Review by PRAkTIKal24 — CHANGES_REQUESTED** [review]
Hi! super nice work! I've left multiple comments in-line and otherwise - please address them in logically grouped commits :)
